src/opera.jl

```diff
@@ -1,72 +1,262 @@
+# Execute instantious interactions
+function execute_instantious_interaction!(opera::Opera)
+    while !isempty(opera.instantious_interactions)
+        call(opera, pop!(opera.instantious_interactions).call)
+    end
+end
+
+"""
+    poke(agent, priority=0)
+Schedule an interaction. Interactions are implemented within an instance `Opera`, sorted by their priorities.
+Internally, reduces to `_interact!(agent)`.
+
+See also [`Opera`](@ref).
+
+# Examples
+```julia
+poke(agent, 1.)
+```
+"""
+function poke(agent, priority = 0.0)
+    add_instantious_interaction!(getopera(agent),
+                                 (; call = () -> _interact!(agent),
+                                  priority = Float64(priority)))
+end
+
+"""
+    @call agent call priority=0
+    @call opera call priority=0
+Schedule an interaction (call). Interactions are implemented within an instance `Opera`, sorted by their priorities.
+Internally, the `call` expression will be transformed to an anonymous function `() -> call`.
+
+See also [`Opera`](@ref).
+
+# Examples
+```julia
+bob_agent = only(getagent(agent, r"bob"))
+@call agent wake_up(bob_agent)
+```
+"""
+macro call(opera, call, priority = 0.0)
+    quote
+        opera = $(esc(opera)) isa Opera ? $(esc(opera)) : getopera($(esc(opera)))
+        add_instantious_interaction!(opera,
+                                     (; call = () -> $(esc(call)),
+                                      priority = Float64($(esc(priority)))))
+    end
 end
 
-"Execute an algebraic interaction."
-function execute_action!(::Opera, ::AbstractOperaCall) end
+"""
+    add_future!(opera, time, call[, id])
+    add_future!(agent, time, call[, id])
+Schedule a (delayed) execution of `call` at `time`. Optionally, provide a textual identifier `id` of the action.
 
-function execute_action!(::Opera, call::AgentCall)
-    if isnothing(call.call)
-        _interact!(call.agent)
+See also [`Opera`](@ref).
+"""
+function add_future! end
+
+function add_future!(opera::Opera, time, call,
+                     id = "future__" * randstring(4))
+    new_action = (; id, call, time)
+    # sorted insert
+    pushfirst!(opera.futures, new_action)
+    ix = 1
+    while ix < length(opera.futures)
+        if new_action.time > opera.futures[ix + 1].time
+            opera.futures[ix] = opera.futures[ix + 1]
+            opera.futures[ix + 1] = new_action
+            ix += 1
+        else
+            break
+        end
+    end
+end
+
+function add_future!(agent::AbstractAlgebraicAgent, args...)
+    add_future!(getopera(agent), args...)
+end
+
+"""
+    @future opera time call [id]
+    @future agent time call [id]
+Schedule a (delayed) execution of `call` at `time`. Optionally, provide a textual identifier `id` of the action.
+
+`call` is an expression, which will be wrapped into an anonymous, parameterless function `() -> call`.
+
+See also [`@future`](@ref) and [`Opera`](@ref).
+"""
+macro future(opera, time, call, id = "future__" * randstring(4))
```

Comment:
What additional functionality does the macro provide? Again, just concerned about long term maintainability.